src/main/java/core/basesyntax/service/impl/ReportProviderImpl.java

```diff
@@ -0,0 +1,27 @@
+package core.basesyntax.service.impl;
+
+import core.basesyntax.model.StorageDao;
+import core.basesyntax.service.ReportProvider;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+public class ReportProviderImpl implements ReportProvider {
+    public static final String REPORT_TITLE = "fruit,quantity";
+    public static final String COLUMN_SEPARATOR = ",";
+    public static final String ROW_SEPARATOR = System.lineSeparator();
+
+    private final Consumer<Integer> positiveValueCheck = operationValue -> {
+        if (operationValue < 0) {
+            throw new RuntimeException("Operation value cannot be negative");
+        }
+    };
+
+    @Override
+    public String provide() {
+        String header = REPORT_TITLE + ROW_SEPARATOR;
+        return header + StorageDao.storage.entrySet().stream()
+                .peek(trans -> positiveValueCheck.accept(trans.getValue()))
+                .map(trans -> trans.getKey() + COLUMN_SEPARATOR + trans.getValue())
+                .collect(Collectors.joining(ROW_SEPARATOR));

```

Comment:
you are still using the usual string concat in 2 places. I was talking about that:

![image](https://github.com/mate-academy/jv-fruit-shop/assets/80981426/5235e07e-3dfe-40f5-b73b-947cbddd2e95)

https://www.geeksforgeeks.org/java-8-streams-collectors-joining-method-with-examples/ - screenshot of the last example here

not sure what to do with the column separator, and how to do that in the best possible way, but we can definetly set `header` as a prefix in `Collectors.joining()`. And now that I thing about this column separator - maybe your `StringBuilder` approach was actually better, but now you know about the `joining()` 😅